### DISCUSSION TITLE
Documentation to create Linkerd extensions

### DISCUSSION THREAD

**Discussion by dashingsat** [linkerd/linkerd2#8679] [Help]
Hello everyone,

We are extremely new to Linkerd. I was trying to see if there are some existing _getting started_ documentations to create own extensions. 

Thanks and Regards
Satya

**Comment by mateiidavid** [comment]
Hi @dashingsat, thanks for the interest! 

We unfortunately do not have any concrete documentation on creating your own extension. Extensions in general make use of the Linkerd API. For example, different viz CLI commands might interact with the metrics api to get the relevant metric:

https://github.com/linkerd/linkerd2/blob/b8e549e239437b8b1150170a5bda27c36699ab8b/viz/metrics-api/proto/viz.proto#L195-L205

Generally, once installed, extensions have a special label `linkerd.io/extension: <extension-name>`. Depending on what you want to do, creating an extension might roughly follow these steps:

1. Create your extension;
2. Create a CLI tool for the extension (if or when applicable);
3. Create a check command and an install/uninstall command in the CLI.

[tapshark](https://github.com/adleong/tapshark) is a great example of an extension that interacts with the tap api to list events. I'd look through the code there as a starting point :)

**Reply by dashingsat** [reply]
Hey @mateiidavid , Thanks for the pointers.. I would explore on this aspect more.

**Reply by dashingsat** [reply]
Hi @mateiidavid, Another very basic question.. Do you have some resources helping contributors to set up a DEV environment with IDE setup and all to debug the changes to LinkerD Proxy..

**Reply by kleimkuhler** [reply]
The easiest way to setup a development environment for the proxy is using its [devcontainer](https://github.com/linkerd/linkerd2-proxy/tree/main/.devcontainer) configuration. Using VS Code (and its devcontainer extension), you should be able to get up and running without any additional configuration; this is how most of the Linkerd maintainers work on the proxy and the control plane.
